docs/design/gluster-block-provisioning.md

```diff
@@ -0,0 +1,350 @@
+# Gluster-Block Provisioning End-To-End Design
+
+## Motivation
+
+While GlusterFS volumes can be used for RWO as well as RWX (and ROX) volumes,
+for some workloads, specifically metadata-heavy, the RWO support is not
+satisfactory with respect to performance. Furthermore, the scalability
+in number of volumes is currently not very good.
+
```

Comment:
@wattsteve ok, now we are getting closer. I think we (I) may have a confusion of terminology. And I need to sort it out:

* For me, coming from storage, a distributed/shared file system like GlusterFS that is typically backing gluster, *does* have locking semantics that makes concurrent access from multiple writers possible. I.e. these distributed locking semantics are what makes RWX access safe.

* Different kinds of storage, like a local block device or local file system that might be a default candidate for RWO volumes do NOT have any locking semantics that prevent multiple mounts accessing it simultaneously from corrupting their data.

So it seems that with "locking semantics" you mean a mechanism that prevents the storage from being used more than one mount/writer or host. If that is the case, then (as I asked in my previous post) this protection mechanism would usually have to be enforced by the mount plugin of kubernetes, since the standard providers do explicitly LACK protection mechanisms. I think I get your point that due to the RWO flag, the users expect to not have to be careful when accessing the file system, since they assume they are alone.

(What by the way is the relevance of locking to a host instead of restricting to a single mount? I think this is a misconception that was introduced by the implementation detail that in most implementations PV mounts are done by a mount on the host and bind-mount into the container, hence all mounts of a PV on a host can be shared by a single host-mount via multiple bind-mounts. But that will still lead to data corruption due to unawareness, and hence we fixed the iSCSI plugin to lock to a single mount...)

So, coming back to the original point: This locking-to-a-mount could easily be achieved by the GlusterFS mount plugin, similarly to what is done in the iSCSI mount plugin. So, if it were only about those protection mechanisms, we would not be having this discussion.

Hence I stand by my point: The main point is performance.

What am I missing?